homeassistant/components/dyson/air_quality.py

```diff
@@ -0,0 +1,146 @@
+"""
+Support for Dyson Pure Cool Air Quality Sensors.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/dyson/
+"""
+import logging
+
+from homeassistant.components.air_quality import (
+    AirQualityEntity, ATTR_ATTRIBUTION, ATTR_NO2, ATTR_PM_10, ATTR_PM_2_5)
+from homeassistant.components.air_quality import DOMAIN as AIQ_DOMAIN
+from homeassistant.components.dyson import DYSON_DEVICES, DOMAIN
+
+DEPENDENCIES = ['dyson']
+
+ATTRIBUTION = 'Dyson purifier air quality sensor'
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_VOC = 'volatile_organic_compounds'
+
+PROP_TO_ATTR = {
+    'attribution': ATTR_ATTRIBUTION,
+    'nitrogen_dioxide': ATTR_NO2,
+    'particulate_matter_10': ATTR_PM_10,
+    'particulate_matter_2_5': ATTR_PM_2_5,
+    'volatile_organic_compounds': ATTR_VOC,
+}
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Set up the Dyson Sensors."""
+    from libpurecool.dyson_pure_cool import DysonPureCool
+
+    _LOGGER.debug("Creating new Dyson air quality sensors")
+    device_serials = []
+    devices = []
+
+    if discovery_info is None:
+        return
+
+    if AIQ_DOMAIN in hass.data:
+        device_serials = {entity.device_serial for entity
+                          in hass.data[AIQ_DOMAIN].entities
+                          if entity.platform.platform_name == DOMAIN}
+
+    # Get Dyson Devices from parent component
+    for device in hass.data[DYSON_DEVICES]:
+        if isinstance(device, DysonPureCool) and \
+                device.serial not in device_serials:
+            devices.append(DysonAirSensor(device))
+    add_entities(devices)
+
+
+class DysonAirSensor(AirQualityEntity):
+    """Representation of a generic Dyson air quality sensor."""
+
+    def __init__(self, device):
+        """Create a new generic air quality Dyson sensor."""
+        self._device = device
+        self._old_value = None
+        self._name = device.name
+
+    async def async_added_to_hass(self):
+        """Call when entity is added to hass."""
+        self.hass.async_add_executor_job(
+            self._device.add_message_listener, self.on_message)
+
+    def on_message(self, message):
+        """Handle new messages which are received from the fan."""
+        # Prevent refreshing if not needed
+        if self._old_value is None or self._old_value != self.state:
+            _LOGGER.debug("Message received for %s device: %s", self.name,
```

Comment:
Should we move the debug log out of the check so that we always log when the callback is called?